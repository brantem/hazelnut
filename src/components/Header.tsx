import { Fragment, useMemo } from 'react';
import { useRouter } from 'next/router';
import Head from 'next/head';

import Button from 'components/Button';

import { Navigation } from 'types/shared';

type Action = { skip?: boolean } & (
  | { children: React.ReactNode; className?: string; onClick: () => void; testId?: string }
  | { render: () => React.ReactNode }
);

export type HeaderProps = {
  navigations: Navigation[];
  actions?: Action[];
};

const Header = ({ navigations, actions }: HeaderProps) => {
  const router = useRouter();
  const navigation = useMemo(() => {
    return navigations.find((navigation) => navigation.href === router.pathname)!;
  }, [navigations, router.pathname]);
  const title = `${navigation.text} - hazelnut`;

  return (
    <>
      <Head>
        <title>{title}</title>
      </Head>
      <header className="flex items-center justify-between py-3 px-4">
        <div className="flex items-center text-xl font-semibold dark:text-white md:text-2xl">
          <h1>{navigation.text}</h1>
        </div>

        {actions?.length ? (
          <div className="flex items-center space-x-2" data-testid="header-actions">
            {actions.map((action, i) => {
              if (action.skip) return null;
              return 'render' in action ? (
                <Fragment key={i}>{action.render()}</Fragment>
              ) : (
                <Button
                  key={i}
                  variant="ghost"
                  size="sm"
                  className={action.className}
                  onClick={action.onClick}
                  data-testid={action.testId}
                >
                  {action.children}
                </Button>
              );
            })}
          </div>
        ) : null}
      </header>
    </>
  );
};

export default Header;
