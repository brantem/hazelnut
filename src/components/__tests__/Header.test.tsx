import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import Header from 'components/Header';

vi.mock('next/router', () => ({
  useRouter() {
    return {
      pathname: '/',
    };
  },
}));

test('Header', () => {
  const onActionClick = vi.fn(() => {});
  const { container } = render(
    <Header
      navigations={[
        { icon: <span />, href: '/', text: 'A' },
        { icon: <span />, href: '/b', text: 'B' },
      ]}
      actions={[
        {
          children: 'Action 1',
          onClick: onActionClick,
        },
        {
          children: 'Action 2',
          onClick: () => {},
          skip: true,
        },
      ]}
    />,
  );

  expect(container).toMatchSnapshot();

  expect(screen.getByText('A')).toBeInTheDocument();
  expect(screen.queryByText('B')).not.toBeInTheDocument();

  screen.getByText('Action 1').click();
  expect(onActionClick).toHaveBeenCalled();
  expect(screen.queryByText('Action 2')).not.toBeInTheDocument();
});
