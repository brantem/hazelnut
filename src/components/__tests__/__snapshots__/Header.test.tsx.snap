// Vitest Snapshot v1

exports[`Header 1`] = `
<div>
  <header
    class="flex items-center justify-between py-3 px-4"
  >
    <div
      class="flex items-center text-2xl font-semibold"
    >
      <h1>
        A
      </h1>
    </div>
    <div
      class="flex items-center space-x-2"
    >
      <button
        class="rounded-full focus:outline-none focus:ring-2 focus:ring-neutral-500 disabled:opacity-70 enabled:hover:bg-neutral-100 px-3 py-1.5 text-sm"
      >
        Action 1
      </button>
      <span>
        Action 3
      </span>
    </div>
  </header>
</div>
`;
